libtransmission/torrent.cc

```diff
@@ -111,6 +111,30 @@ tr_torrent* tr_torrentFindFromObfuscatedHash(tr_session* session, tr_sha1_digest
     return nullptr;
 }
 
+bool tr_torrentSetMetainfoFromFileIfMagnet(tr_torrent* tor, tr_torrent_metainfo* metainfo, char const* filename)
+{
+    if (tr_torrentHasMetadata(tor))
+    {
+        return false;
+    }
+    tr_error* error = nullptr;
+
+    tr_torrentUseMetainfoFromFile(tor, metainfo, filename, &error);
+
+    if (error != nullptr)
+    {
+        tor->setLocalError(fmt::format(
+            _("Couldn't copy '{path_in}' to '{path_out}': {error} ({error_code})"),
+            fmt::arg("path_in", filename),
+            fmt::arg("path_out", tor->torrentFile()),
+            fmt::arg("error", error->message),
+            fmt::arg("error_code", error->code)));
+        tr_error_clear(&error);
```

Comment:
This codepath returns `true` even if the operation failed, which is a little weird.

I see the point, since the return value is "was this a magnet link" rather than "did the operation succeed".  But then there's no way to know if the operation actually succeeded...

Maybe this should be renamed `tr_torrentSetMetainfoFromFile()`, return false if either (a) it already has metainfo or (b) UseMetainfoFromFile()  fails, and Controller.mm should use something like

```
if (!tr_torrentHasMetadata(foundTorrent))
{
    notify dupe error
}
else if (!tr_torrentSetMetainfoFromFile(foundTorrent, ...))
{
    notify unable to add metainfo
}
```